virtio-queue/src/queue/verification.rs

```diff
@@ -565,3 +565,139 @@ fn verify_add_used() {
+// the provided ordering. Takes used_ring address and queue size directly.
+fn get_avail_event<M: GuestMemory>(
+    used_ring_addr: GuestAddress,
+    queue_size: u16,
+    mem: &M,
+    order: Ordering,
+) -> Result<u16, Error> {
+    // This can not overflow an u64 since it is working with relatively small numbers compar
+    // to u64::MAX.
+    let avail_event_offset =
+        VIRTQ_USED_RING_HEADER_SIZE + VIRTQ_USED_ELEMENT_SIZE * u64::from(queue_size);
+    let addr = used_ring_addr
+        .checked_add(avail_event_offset)
+        .ok_or(Error::AddressOverflow)?;
+
+    mem.load::<u16>(addr, order).map_err(Error::GuestMemory)
+}
+
+// Get the value of the `flags` field of the used ring, applying the specified ordering.
+fn get_used_flags<M: GuestMemory>(queue: &Queue, mem: &M, order: Ordering) -> Result<u16, Error> {
+    mem.load::<u16>(queue.used_ring, order)
+        .map(u16::from_le)
+        .map_err(Error::GuestMemory)
+}
+/// # Specification (VirtIO 1.3, Section 2.7.7: "Used Buffer Notification Suppression")
+///
+/// This proof checks:
+/// - If event_idx is not enabled, `set_notification(false)` sets used.flags to 1 (NO_NOTIFY).
+/// - If event_idx is enabled, the call is a no-op.
+#[kani::proof]
+#[kani::unwind(0)]
+fn verify_set_notification_true() {
+    let ProofContext { mut queue, memory } = kani::any();
+    if queue
+        .set_notification(&memory, true /* enable notification */)
+        .is_ok()
+    {
+        if queue.event_idx_enabled {
+            // Since VIRTIO_F_EVENT_IDX is negotiated, we make sure that set_notification
+            // has updated the used.avail_event field with the tail position of
+            // the avail ring.
+            let used_ring_addr = queue.used_ring;
+            let queue_size = queue.size();
+            kani::cover!();
+            assert_eq!(
+                get_avail_event(used_ring_addr, queue_size, &memory, Ordering::Relaxed).unwrap(),
+                queue.next_avail.0
+            );
+        } else {
+            // If VIRTIO_F_EVENT_IDX is not negotiated, we make sure that the
+            // used.flags field is set to 0, meaning that the driver should not
+            // send notifications to the device.
+            kani::cover!();
+            assert_eq!(
+                get_used_flags(&queue, &memory, Ordering::Relaxed).unwrap(),
+                0
+            );
+        }
+    }
+}
+
+/// # Specification (VirtIO 1.3, Section 2.7.7: "Used Buffer Notification Suppression")
+///
+/// This proof checks:
+/// - If event_idx is not enabled, `set_notification(false)` sets used.flags to 1 (NO_NOTIFY).
+/// - If event_idx is enabled, the call is a no-op.
+#[kani::proof]
+#[kani::unwind(0)]
+fn verify_set_notification_false() {
+    let ProofContext { mut queue, memory } = kani::any();
+    let result = queue.set_notification(&memory, false /* disable notification */);
+    if !queue.event_idx_enabled {
+        // Check for Sec 2.7.10
```

Comment:
`Sec`, should be `Spec` right?

Also the test doc talks about `Specification (VirtIO 1.3, Section 2.7.7: "Used Buffer Notification Suppression")` what 2.7.10 is? Is related to virtio 1.3?

Please be more verbose with that references.

Reply:
I think here you could add that Section 2.7.10 of the Virtio spec explains how the device can indicates to the driver that notifications are not required when adding buffers into the avail right.